Hydra-node architecture

v0d1ch opened the pull request:
## Why
We want to describe the inner workings of a hydra-node a bit better for the users thus this document presented as the readme in the hydra-node package.



---

<!-- Tick off or strike-through / remove if not applicable -->
* [ ] CHANGELOG updated
* [x] Documentation updated
* [ ] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Hydra-node architecture

ch1bo committed:
Create a dummy hydra-spec.pdf in docs for writing

As a writer of the docs, I might not want to build the specification and
so the 'yarn start' workflow puts a dummy specification in place of the
hydra-spec.pdf. This is needed as we import the file to cache bust URL.

ffakenz reviewed (commented):
does it worth to mention the ADR section here?

ffakenz reviewed (commented):
idk if makes sense but usually this doc pages come along with many pictures.

not saying the architecture-diagram is not enough, but maybe describe the main handlers, components and types on each layer, or just mention them and add references to the haddock, would be cool imo.

abailly committed:
Upgrade docusaurus to 2.14.0

abailly committed:
Draft C4 component diagram for architecture

Support for this kind of diagram in Mermaid is not great, seems like
the resulting diagram will have a high probability of ending up being
ugly...

ghost commented:
Here is how the C4 diagram looks like using inline mermaid. Using [C4 PlantUML](https://github.com/plantuml-stdlib/C4-PlantUML) I was able to create [much better looking](https://input-output-hk.github.io/ouroboros-consensus/haddocks/index.html) diagrams by forcing the relative positioning of components but this option does not seem available in mermaid. 

I will check if there's a build time solution we could embed in CI.

<img width="1197" alt="Screenshot 2023-04-29 at 09 26 47" src="https://user-images.githubusercontent.com/79840582/235289821-cfcf6d5f-8380-493f-9c9e-8e49e5b43e9a.png">

abailly committed:
Add more components to architecture diagram

abailly committed:
Make architecture a section in itself and polish C4 diagram

* Link to relevant section of the code from the diagram
* Turn 'Architecture' into a toplevel section with more details for
the various parts underneath it (currently only networking)

abailly committed:
Pin upper bound for asyncapi/specs package version

Version 4.3.3 if @asyncapi/specs package is borked.

abailly committed:
Fix broken link

abailly committed:
Fix japanese pages following upgrade to docusaurus

abailly committed:
Use yarn resolutions mechanism to fix broken asyncapi specs release

abailly committed:
Protect static haddock html files from SPA routing

Trying the solution from https://docusaurus.io/docs/advanced/routing#escaping-from-spa-redirects
this should fix #845 but it does not seem to work correctly locally with yarn serve

abailly committed:
Add instructions to recreate the diagram

ffakenz commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+
+Container(hydra_node_peer, "Hydra Node (peer)", "Local cadano-node")
+Container(cardano_node, "Cardano Node", "Local cadano-node")
+Container(cardano_node_peer, "Cardano Node (peer)", "Local cadano-node")
```
```suggestion
Container(cardano_node_peer, "Cardano Node (peer)", "Local cardano-node")
```

ffakenz commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+   Rel_L(chain, wallet, "Sign transactions", "")
+}
+
+Container(hydra_node_peer, "Hydra Node (peer)", "Local cadano-node")
```
```suggestion
Container(hydra_node_peer, "Hydra Node (peer)", "Local cardano-node")
```

ffakenz commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+#### Tiny Wallet
+
+Hydra also maintains an internal "wallet" which is currently part of the Chain components layer, to handle the payment of transaction fees and signing transctions using a dedicated Cardano signing key. There are [plans](https://github.com/input-output-hk/hydra/issues/215) to make it possible to use an external  wallet to balance and sign Hydra node's transactions.
```
```suggestion
Hydra also maintains an internal "wallet" which is currently part of the Chain components layer, to handle the payment of transaction fees and signing transctions using a dedicated Cardano signing key. There are [plans](https://github.com/input-output-hk/hydra/issues/215) to make it possible to use an external wallet to balance and sign Hydra node's transactions.
```

ffakenz commented on docs/core-concepts/index.md:
````diff
@@ -15,15 +15,15 @@ There are different flavors and extensions of the Hydra Head protocol, but let's
 ```mdx-code-block
 import DocCardList from '@theme/DocCardList';
-import {useDocsSidebar} from '@docusaurus/theme-common';
+import {useDocsSidebar} from "@docusaurus/theme-common/internal";
````
```suggestion
import {useDocsSidebar} from '@docusaurus/theme-common/internal';
```

ffakenz replied:
jaja do not hate me.

ffakenz commented on docs/package.json:
```diff
@@ -16,19 +16,25 @@
     "write-heading-ids": "docusaurus write-heading-ids",
-    "enrich-document-metadata": "node ./scripts/enrich-document-metadata.js"
+    "enrich-document-metadata": "node ./scripts/enrich-document-metadata.js",
+    "dummy-spec": "echo 'Do a nix build .#spec and put this instead here' > static/hydra-spec.pdf"
```
what is this for?

ffakenz commented on docs/docusaurus.config.js:
```diff
@@ -279,6 +279,13 @@ const config = {
+    mermaid: true,
+  },
+
+  themes: ['@docusaurus/theme-mermaid'],
```
🧜‍♀️ 🧜‍♂️

ffakenz commented on docs/docs/haskell_packages.md:
```diff
@@ -4,23 +4,23 @@ sidebar_position: 99
-| [plutus-cbor](/haddock/plutus-cbor/index.html)               | Implementation of CBOR encoders, compatible with on-chain Plutus validators.        |
-| [hydra-prelude](/haddock/hydra-prelude/index.html)           | Custom Hydra Prelude used across other Hydra packages.                              |
-| [hydra-cardano-api](/haddock/hydra-cardano-api/index.html)   | A wrapper around the `cardano-api`, with era-specialized types and extra utilities. |
+| [plutus-merkle-tree](pathname:///haddock/plutus-merkle-tree/index.html) | Implementation of Merkle Trees, compatible with on-chain Plutus validators.         |
```
[interesting](https://docusaurus.io/docs/advanced/routing#escaping-from-spa-redirects) 🤔

ffakenz commented on docs/core-concepts/architecture/hydra-components.svg:
🧑‍🎨

Commit:
Update docs/core-concepts/architecture/architecture-c4.puml

Co-authored-by: Franco Testagrossa <franco.testagrossa@iohk.io>

Commit:
Update docs/core-concepts/architecture/architecture-c4.puml

Co-authored-by: Franco Testagrossa <franco.testagrossa@iohk.io>

Commit:
Update docs/core-concepts/architecture/index.md

Co-authored-by: Franco Testagrossa <franco.testagrossa@iohk.io>

Commit:
Update docs/core-concepts/index.md

Co-authored-by: Franco Testagrossa <franco.testagrossa@iohk.io>

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+   Component(persistence, "Persistence", "JSON/FS", "Persists state of the head on disk", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Persistence.hs")
+   Component(head_logic, "Head Logic", "", "Maintains internal Head state and processes events", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/HeadLogic.hs")
+   Component(api_server, "API Server", "", "Dispatch client commands & queries and serves outputs", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/API/Server.hs")
+   Component(chain, "Direct Chain", "", "Connection to Cardano\nObserve txs on-chain and post tx to chain", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Chain/Direct.hs")
```
Should: Either call the network layer `Ouroboros Network` or this one simply `Chain` layer.

Not sure about the notation, but I personally would prefer to have simple names about the components and maybe detail links to the actual implementation in the text. For example, link to `Chain.hs` here and mention that the actual implementation located in `Direct.hs` uses a direct connection to a `cardano-node` in the supporting text.

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+   Component(api_server, "API Server", "", "Dispatch client commands & queries and serves outputs", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/API/Server.hs")
+   Component(chain, "Direct Chain", "", "Connection to Cardano\nObserve txs on-chain and post tx to chain", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Chain/Direct.hs")
+   Component(wallet, "Tiny Wallet", "", "Embedded Wallet to sign transactions and pay fees", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Chain/Direct/Wallet.hs")
+   Component(network, "Network", "", "Manages connections to peers", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Network/Ouroboros.hs")
```
Could: Link to the interface module `Network.hs`

Not sure if we want to be specific or generic. The level of detail of the graph suggests a generic component diagram.

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+AddElementTag("external", $bgColor="#23aa12")
+
+Container_Boundary(hydra_node, "Hydra Node") {
+   Component(persistence, "Persistence", "JSON/FS", "Persists state of the head on disk", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Persistence.hs")
```
Can we make this non-connected like `Logging` and `Monitoring` blocks? That would underline better that this is in fact a bit of a cross-cutting concern (otherwise we would need to have a line between API and persistence to be correct).

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+Container_Boundary(hydra_node, "Hydra Node") {
+   Component(persistence, "Persistence", "JSON/FS", "Persists state of the head on disk", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Persistence.hs")
+   Component(head_logic, "Head Logic", "", "Maintains internal Head state and processes events", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/HeadLogic.hs")
+   Component(api_server, "API Server", "", "Dispatch client commands & queries and serves outputs", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/API/Server.hs")
```
```suggestion
   Component(api_server, "API Server", "", "Handle client commands & queries and serves outputs", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/API/Server.hs")
```

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+   Component(head_logic, "Head Logic", "", "Maintains internal Head state and processes events", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/HeadLogic.hs")
+   Component(api_server, "API Server", "", "Dispatch client commands & queries and serves outputs", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/API/Server.hs")
+   Component(chain, "Direct Chain", "", "Connection to Cardano\nObserve txs on-chain and post tx to chain", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Chain/Direct.hs")
+   Component(wallet, "Tiny Wallet", "", "Embedded Wallet to sign transactions and pay fees", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Chain/Direct/Wallet.hs")
```
```suggestion
   Component(wallet, "Wallet", "", "Embedded Wallet to sign transactions and pay fees", $link="https://github.com/input-output-hk/hydra/blob/master/hydra-node/src/Hydra/Chain/Direct/Wallet.hs")
```

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+   Component(ledger, "Cardano Ledger", "", "Maintains Layer 2 State", $tags="external")
+
+   BiRel_L(head_logic, network, "Hydra Protocol Messages", "")
+   BiRel_R(head_logic, api_server, "Input & Output", "")
```
Should: be a bit more specific.. at least refer to the types used here.

```suggestion
   BiRel_R(head_logic, api_server, "ClientInput & ServerOutput", "")
```

Or even better label one arrow in each direction with the respective label?

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
Could: link to the code or the haddock?

We rarely do have any `README.md` next to the `.hs` files, but sometimes we do have module documentation? Also, IMO this page could be another entry point to our eventually great haddocks. I would lean to linking the haddock files

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+Container(cardano_node_peer, "Cardano Node (peer)", "Local cardano-node")
+Container(hydra_client, "Hydra Client", "")
+Component(smart_contracts, "Hydra Smart Contracts", "Plutus", "On-chain smart contracts controlling the behaviour of the protocol on-chain", $link="https://github.com/input-output-hk/hydra/tree/master/hydra-plutus")
+Rel(chain, smart_contracts, "References", "")
```
```suggestion
Rel(chain, smart_contracts, "Use", "")
```

Or: `include`, or `compiled in` .. I feel "referencing" is a bit too weak verb.

ch1bo commented on docs/core-concepts/architecture/architecture-c4.puml:
```diff
@@ -0,0 +1,40 @@
+BiRel_D(chain, cardano_node, "", "N2C protocol")
+BiRel_L(cardano_node, cardano_node_peer, "", "N2N protocol")
+BiRel(hydra_node_peer, cardano_node_peer, "", "N2C protocol")
+BiRel_L(network, hydra_node_peer, "", "N2C protocol")
```
```suggestion
BiRel_L(network, hydra_node_peer, "", "Hydra Network protocol")
```

Although that protocol is no-where documented. Certainly not N2C.

ch1bo commented on docs/core-concepts/architecture/hydra-components.svg:
I'm torn.. I like the reproducibility of this diagram, but it also is more weird and probably hard to accomodate spatial information. For example, the fact that cardano-node boxes were drawn below and on the same level should convey the fact that these are representing the layer 1, while hydra-nodes connected to each other are the layer 2.

ch1bo commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+#### Direct Chain
+
+The _Chain_ component is responsible for interfacing the Hydra node with the Cardano (aka. Layer 1) chain. It uses the `Node-to-Client` protocol to both "follow the chain" and observe new blocks and transactions which can change the state of the head, and submit transactions in response to client's requests or as needed to advance the protocol's state. It connects to a locally spun cardano-node u ...
```
```suggestion
The _Chain_ component is responsible for interfacing the Hydra node with the Cardano (aka. Layer 1) chain. It uses the `Node-to-Client` protocol to both "follow the chain" and observe new blocks and transactions which can change the state of the head, and submit transactions in response to client's requests or as needed to advance the protocol's state. It connects to a local cardano-node using a domain socket and contains the _off-chain_ logic, based on cardano-api, that knows how to observe and build transactions relevant to the Hydra protocol.
```

ch1bo commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+:::
+
+#### Tiny Wallet
```
```suggestion
#### Wallet
```

There a single occurrence of `Tiny` in the code.. and it never was tiny. I'd call it just wallet or `Internal Wallet` if you want.

ch1bo commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+#### Tiny Wallet
+
+Hydra also maintains an internal "wallet" which is currently part of the Chain components layer, to handle the payment of transaction fees and signing transctions using a dedicated Cardano signing key. There are [plans](https://github.com/input-output-hk/hydra/issues/215) to make it possible to use an external wallet to balance and sign Hydra node's transactions.
```
The linked plans have nothing to do with balancing and signing Hydra node transactions. I would drop that reference.

ch1bo replied:
```suggestion
The Hydra node maintains an internal wallet using the Cardano signing key provided to the `hydra-node`. This is used to handle the payment of transaction fees and signing transactions.
```

ch1bo commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+### Logging
+
+The Hydra node logs all side-effects occuring internally as JSON-formatted messages to the _standard output_ stream attached to its process. The format of the logs is documented as a [JSON schema](https://raw.githubusercontent.com/input-output-hk/hydra/master/hydra-node/json-schemas/logs.yaml), and follows the principles outlined in [ADR-9](/adr/9).
```
```suggestion
The Hydra node logs all side-effects occurring internally as JSON-formatted messages to the _standard output_ stream attached to its process. The format of the logs is documented as a [JSON schema](https://raw.githubusercontent.com/input-output-hk/hydra/master/hydra-node/json-schemas/logs.yaml), and follows the principles outlined in [ADR-9](/adr/9).
```

ch1bo commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+### Monitoring
+
+The Hydra node [optionally](http://localhost:3000/head-protocol/docs/getting-started/quickstart#hydra-node-options) exposes [Prometheus](https://prometheus.io/)-compliant _metrics_ through an HTTP server, on the standard `/metrics` endpoint.
```
```suggestion
The Hydra node [optionally](https://hydra.family/head-protocol/docs/getting-started/quickstart#hydra-node-options) exposes [Prometheus](https://prometheus.io/)-compliant _metrics_ through an HTTP server, on the standard `/metrics` endpoint.
```

ch1bo commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+This "component" represents  all of the Hydra smart contracts needed for Head protocol operation. Currently the contracts are written using `Plutus-Tx`. The scripts are optimized using custom `ScriptContext` and error codes for now.
+
+### API layer
```
```suggestion
### API
```

ch1bo commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+### Chain Interaction
+
+#### Direct Chain
```
```suggestion
### Chain
```

ch1bo commented on docs/core-concepts/architecture/index.md:
```diff
@@ -0,0 +1,75 @@
+
+:::info
+
+The reason why this component is called _Direct_ chain is historical: Initially, we thought of using the [Plutus Application Backend](https://plutus-apps.readthedocs.io/en/latest/plutus/explanations/pab.html) to handle interaction between the off- and on-chain parts of the Hydra protocol but we switched to use the lower level node-to-client protocol.
```
Not sure how important this anecdote / historic information is here. We could still use the PAB (or what remains of it), or we could use something completely different.

I would drop it.

ch1bo reviewed (approved):
Good start for an architecture page. I'm not sure about the C4 diagram though and it's generative nature. As spelled out in a comment I like to use spatial distribution of boxes to convey information.

Nonetheless I can approve this, there are minor things to fix though.

One more thing: I think this should be removed now: https://hydra.family/head-protocol/haddock/hydra-node/Hydra-Node.html

Commit:
Update docs/core-concepts/architecture/architecture-c4.puml

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

Commit:
Update docs/core-concepts/architecture/architecture-c4.puml

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

Commit:
Update docs/core-concepts/architecture/index.md

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

Commit:
Update docs/core-concepts/architecture/index.md

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

Commit:
Update docs/core-concepts/architecture/index.md

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

abailly committed:
Use more generic components in architecture diagram

abailly committed:
Remove architecture diagram from source code.

abailly committed:
Fix link from ADR

ch1bo commented on docs/adr/2021-10-23_010-use-direct-chain.md:
```diff
@@ -30,6 +30,4 @@ For the time being, until _Plutus Application Framework_ is released, we will im
-* An updated architecture diagram:
-
-![](../../hydra-node/images/hydra-architecture-direct.jpg)
+* See [Architecture page](/core-concepts/architecture) for details
```
```suggestion
```

Would drop it, as it was pointing to a picture that does not exist and it's not a value-adding consequence. An ADR always may result in a changed architecture :shrug:

Commit:
Update docs/adr/2021-10-23_010-use-direct-chain.md

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ghost commented:
Can we merge this?

ch1bo commented:
> Can we merge this?

@abailly-iohk was waiting for you to address or discard my comments like this one: https://github.com/input-output-hk/hydra/pull/838/files#r1188250333

ch1bo committed:
Apply suggestions from code review

ch1bo committed:
Create a script to regenerate plantuml files

ch1bo committed:
Regenerate plantuml drawings

ghost commented:
I have discarded them.

ch1bo committed:
Make regenerate-plantuml step in docs optional

The CI currently has no dependencies installed and it's more of a best
effort right now.
